ocaml/xs-trace/xs_trace.ml

```diff
@@ -12,48 +12,94 @@
-      let submit_json json =
-        if json <> "" then
-          let result = Tracing_export.Destination.Http.export ~url json in
-          match result with
-          | Ok _ ->
-              ()
-          | Error err ->
-              Printf.eprintf "Error: %s" (Printexc.to_string err) ;
-              exit 1
-      in
-      let rec export_file orig =
-        if Sys.is_directory orig then
-          let files = Sys.readdir orig in
-          let file_paths = Array.map (Filename.concat orig) files in
-          Array.iter export_file file_paths
-        else if Filename.check_suffix orig ".zst" then
-          Xapi_stdext_unix.Unixext.with_file orig [O_RDONLY] 0o000
-          @@ fun compressed_file ->
-          Zstd.Fast.decompress_passive compressed_file @@ fun decompressed ->
-          if Filename.check_suffix orig ".ndjson.zst" then
+open Cmdliner
+
+module Exporter = struct
+  module Unixext = Xapi_stdext_unix.Unixext
+
+  let ( >> ) f g x = g (f x)
+
+  (* Submit JSON to a specified endpoint. *)
+  let submit_json url json =
+    if json <> "" then
+      match Tracing_export.Destination.Http.export ~url json with
+      | Error err ->
+          Printf.eprintf "Error: %s" (Printexc.to_string err) ;
+          exit 1
+      | _ ->
+          ()
+
+  (** Export traces from file system to a remote endpoint. *)
+  let export erase src dst =
+    let dst = Uri.of_string dst in
+    let submit_json = submit_json dst in
+    let open Filename in
+    let rec export_file = function
+      | path when Sys.is_directory path ->
+          (* Recursively export trace files. *)
+          Sys.readdir path |> Array.iter (concat path >> export_file)
+      | path when check_suffix path ".zst" ->
+          (* Decompress compressed trace file and decide whether to
+             treat it as line-delimited or not. *)
+          let ( let* ) = ( @@ ) in
+          let* compressed = Unixext.with_file path [O_RDONLY] 0o000 in
+          let* decompressed = Zstd.Fast.decompress_passive compressed in
+          if check_suffix path ".ndjson.zst" then
             let ic = Unix.in_channel_of_descr decompressed in
-            Xapi_stdext_unix.Unixext.lines_iter
-              (fun line -> submit_json line)
-              ic
+            Unixext.lines_iter submit_json ic
           else
-            let json = Xapi_stdext_unix.Unixext.string_of_fd decompressed in
+            let json = Unixext.string_of_fd decompressed in
             submit_json json
-        else if Filename.check_suffix orig ".ndjson" then
-          Xapi_stdext_unix.Unixext.readfile_line
-            (fun line -> submit_json line)
-            orig
-        else
-          let json = Xapi_stdext_unix.Unixext.string_of_file orig in
+      | path when check_suffix path ".ndjson" ->
+          (* Submit traces line by line. *)
+          Unixext.readfile_line submit_json path
+      | path ->
+          (* Assume any other extension is a valid JSON file. *)
```

Comment:
Do we really want this? We are passing an arbitrary file to the endpoint. Why are we passing some files by line and others in one go? I would suggest to limit this at least to `.json` even when not validating the input.